### PR TITLE
Add tunterm acl plugin support.

### PR DESCRIPTION
**Pull request by dypet**
### Overview:
These changes add support for the recently added tunterm_acl vpp plugin (https://github.com/sonic-net/sonic-platform-vpp/pull/114) in sonic-vpp. 

### Changes:
Added API calls to the new plugin in SaiVppXlate.c, and added logic to SwitchStateBaseAcl.cpp so that the new plugin is called with the appropriate fields instead of the regular ACL plugin when SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is received. Ports are recorded in m_acl_tbl_hw_ports_map at table creation so that they may be passed to the tunterm_acl plugin and bound by the new plugin the first time an ACE with SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is configured. tunterm_acl ACLs are tracked in m_tunterm_acl_swindex_map similarly to m_acl_swindex_map for a given tbl_oid.

### CODE REVIEW

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 1]
```diff
@@ -597,65 +898,100 @@ sai_status_t SwitchStateBase::AclTblConfig(
 				meta->attridname);
 	    }
 	    
-	    if (attr->id == SAI_ACL_ENTRY_ATTR_FIELD_ACL_RANGE_TYPE) {
+        if(attr->id == SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED) {
+            if(attr->value.aclfield.data.booldata == true) {
+                update_tunterm_rule = true;
+            }
```
Comment (abdbaig) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:904]: indent this block

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 2]
```diff
@@ -1290,6 +1639,11 @@ sai_status_t SwitchStateBase::aclBindUnbindPorts(
 			is_bind ? "bound to": "unbound from", hwif_name.c_str());
     }
 
+    for(auto hwif: hwif_list) {
+        m_acl_tbl_hw_ports_map[tbl_oid].push_back(hwif);
+    }
+    
+
```
Comment (abdbaig) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1646]: remove additional lines

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c [hunk 3]
```diff
@@ -1828,6 +1890,148 @@ int vpp_acl_del (uint32_t acl_index)
...
+    mp->tunterm_acl_index = htonl(*tunterm_index);
+
+    for (idx = 0; idx < count; idx++) {
+        in_rule = &in_acl->rules[idx];
+        vpp_rule = &mp->r[idx];
+
+        addr = &in_rule->dst_prefix;
+        api_addr = &vpp_rule->dst;
+
+        if (addr->sa_family == AF_INET) {
```
Comment (abdbaig) [inline platform/saivpp/vpplib/vppxlate/SaiVppXlate.c:1981]: Here too you can use vpp_to_vl_api_ip_addr() to convert IP addresses

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 4]
```diff
@@ -489,6 +663,94 @@ static bool cmp_priority (
...
+    if (tunterm_idx_it == m_tunterm_acl_swindex_map.end()) {
+        SWSS_LOG_WARN("No tunterm ACL configured for table %s", sai_serialize_object_id(tbl_oid).c_str());
+        return SAI_STATUS_SUCCESS;
+    }
+    uint32_t tunterm_acl_swindex = tunterm_idx_it->second;
+
+    status = vpp_tunterm_acl_del(tunterm_acl_swindex);
+
+    if (status == SAI_STATUS_SUCCESS) {
+	    m_tunterm_acl_swindex_map.erase(tunterm_idx_it);
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:745]: Fix indentation here and throughout.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 5]
```diff
@@ -1274,11 +1613,21 @@ sai_status_t SwitchStateBase::aclBindUnbindPorts(
...
+        if(ret == 0) {
+            hwif_list.push_back(hwif_name);
+            tunterm_acl_bindunbind(tbl_oid, true, hwif_name);
+        }
+    }
+	else {
 	    ret = vpp_acl_interface_unbind(hwif_name.c_str(), acl_swindex, is_input);
-
+        if(ret == 0) {
+            hwif_list.remove(hwif_name);
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1627]: In the unbind case, wouldn't hwif_list always be empty?

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 6]
```diff
@@ -1290,6 +1639,11 @@ sai_status_t SwitchStateBase::aclBindUnbindPorts(
 			is_bind ? "bound to": "unbound from", hwif_name.c_str());
     }
 
+    for(auto hwif: hwif_list) {
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1642]: Does this correctly update the ports_map in the unbind case?

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 7]
```diff
@@ -563,6 +825,30 @@ sai_status_t SwitchStateBase::AclTblConfig(
...
+    size_t n_entries = 0;
+    size_t n_tunterm_entries  = 0;
+    for (auto ace: ordered_aces) {
+        p_ace = &aces[ace.index];
+        const sai_attribute_t *attr;
+        bool ttf_flag_set = false;
+        for (uint32_t i = 0; i < p_ace->attrs_count; i++) {
+            attr = &p_ace->attrs[i];
+            if(attr->id == SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED) {
+                if(attr->value.aclfield.data.booldata == true) {
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:837]: Collapse these into a single if statement as they are part of the same check.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 8]
```diff
@@ -563,6 +825,30 @@ sai_status_t SwitchStateBase::AclTblConfig(
...
+                    ttf_flag_set=true;
+                }
+                break;
+            }
+        }
+        if(ttf_flag_set) {
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:843]: Where possible, follow formatting convention of the file (eg. add a space after your if)

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.h [hunk 9]
```diff
@@ -81,6 +81,21 @@ extern "C" {
...
+        char hwif_name[64];
+        uint8_t  ip_protocol;
+        vpp_ip_addr_t next_hop_ip;
+    } tunterm_acl_rule_t;
+
+    typedef struct _vpp_tunterm_acl_ {
+        char *acl_name;
+        uint32_t count;
+        tunterm_acl_rule_t rules[0];
+    } vpp_tunerm_acl_t;
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/vppxlate/SaiVppXlate.h:96]: Correct structure name to vpp_tunterm_acl_t

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 10]
```diff
@@ -597,65 +898,100 @@ sai_status_t SwitchStateBase::AclTblConfig(
 				meta->attridname);
 	    }
 	    
-	    if (attr->id == SAI_ACL_ENTRY_ATTR_FIELD_ACL_RANGE_TYPE) {
+        if(attr->id == SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED) {
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:901]: Collapse into 1 if as it's part of the same check.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 11]
```diff
@@ -597,65 +898,100 @@ sai_status_t SwitchStateBase::AclTblConfig(
...
     uint32_t acl_swindex;
-
     auto vpp_idx_it = m_acl_swindex_map.find(tbl_oid);
-    if (vpp_idx_it == m_acl_swindex_map.end()) {
-	acl_swindex = 0;
-	acl_replace = false;
-    } else {
-	acl_swindex = vpp_idx_it->second;
-	acl_replace = true;
+    if(acl->count > 0) {
```
Comment (AkeelAli) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:938]: This block shouldn't have changed, yet it's marked as a diff here due to indentation changes. Please revert this.

**Review by AkeelAli — COMMENTED** [review]
If not already done, can you please re-run the basic ACL redirect tests with the latest plugin (as some changes were made before it was merged). You can wait until the final version of this review to do that.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 12]
```diff
@@ -563,6 +825,30 @@ sai_status_t SwitchStateBase::AclTblConfig(
     ordered_aces.sort(cmp_priority);
     // SWSS_LOG_NOTICE("# aces %u after sort", ordered_aces.size());
 
+    size_t n_entries = 0;
+    size_t n_tunterm_entries  = 0;
+    for (auto ace: ordered_aces) {
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:830]: This function is too long. We'd better break it into a few smaller functions. For example, for the part that converting ACE to acl_tbl_entries_t and sort can be one function. From sorted  acl_tbl_entries_t to vpp_acl_t and vpp_tunerm_acl_t can be another function.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 13]
```diff
@@ -597,65 +898,100 @@ sai_status_t SwitchStateBase::AclTblConfig(
...
+            if(tunterm_acl->count>0 && (tunterm_rule<(&tunterm_acl->rules[0]+n_tunterm_entries))) {
+                status = tunterm_acl_rule_field_update((sai_acl_entry_attr_t) attr->id, &attr->value, tunterm_rule);
+	        }
+        }
 	}
-	rule++;
+    
+    if(update_tunterm_rule) {
+        if(acl->count>0 && (rule<(&acl->rules[0]+n_entries))) {
+            memset(rule, 0, sizeof(*rule)); // clear any fields that were added to regular rule
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:924]: Can you explain this?

File: platform/saivpp/vpplib/SwitchStateBase.h [hunk 14]
```diff
@@ -1016,6 +1018,12 @@ namespace saivpp
 		_In_ sai_object_id_t tbl_oid,
 		_In_ bool is_add);
 
+            sai_status_t tbl_hw_ports_map_delete(
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBase.h:1021]: Please add some documents to explain these new functions

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 15]
```diff
@@ -597,65 +898,100 @@ sai_status_t SwitchStateBase::AclTblConfig(
...
-    }
     SWSS_LOG_NOTICE("ACL table %s %s status %d", tbl_sid.c_str(),
 		    acl_replace ? "replace" : "add", status);
+    
+    if(status == SAI_STATUS_SUCCESS) {
+        status = tunterm_acl_add_replace(tunterm_acl, tbl_oid);
+        SWSS_LOG_NOTICE("Tunterm ACL table %s %s status %d", tbl_sid.c_str(),
+		    acl_replace ? "replace" : "add", status);
+    } else {
+        SWSS_LOG_NOTICE("ACL plugin operation failed, skipping tunterm configuration.");;
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:987]: This should be an error

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 16]
```diff
@@ -405,6 +405,180 @@ sai_status_t acl_rule_field_update(
...
+    }
+
+    std::string hwif_name;
+	if (!vpp_get_hwif_name(port_oid, vlan_id, hwif_name)) {
+	    SWSS_LOG_WARN("VPP hwif name not found for port %s", sai_serialize_object_id(port_oid).c_str());
+        return SAI_STATUS_SUCCESS;
+	}
+
+    switch (ip_address.addr_family) {
+        case SAI_IP_ADDR_FAMILY_IPV4:
```
[the remark this answers is not in the document]
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:471]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 17]
```diff
@@ -405,6 +405,180 @@ sai_status_t acl_rule_field_update(
...
+    _Out_ tunterm_acl_rule_t      *rule)
+{
+    sai_status_t status;
+
+    assert(NULL != value);
+    status = SAI_STATUS_SUCCESS;
+
+    switch (attr_id) {
+    case SAI_ACL_ENTRY_ATTR_FIELD_SRC_IP:
+    case SAI_ACL_ENTRY_ATTR_FIELD_INNER_SRC_IP:
```
[the remark this answers is not in the document]
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 18]
```diff
@@ -405,6 +405,180 @@ sai_status_t acl_rule_field_update(
...
+    if(rule->ip_protocol == 1) {
+        SWSS_LOG_NOTICE("Tunterm acl rule has next-hop IP %x", rule->next_hop_ip.addr.ip4.sin_addr.s_addr);
+    } else if (rule->ip_protocol == 2) {
+        for (int i=0; i<16; i++) {
+            SWSS_LOG_NOTICE("Tunterm acl rule has next-hop IP6 %u", rule->next_hop_ip.addr.ip6.sin6_addr.s6_addr[i]);
+        }
+        
+    }
+    
+
```
[the remark this answers is not in the document]
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 1 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 2 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c (hunk 3 above)
Comment (dypet) [reply platform/saivpp/vpplib/vppxlate/SaiVppXlate.c]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 19]
```diff
@@ -1274,11 +1613,21 @@ sai_status_t SwitchStateBase::aclBindUnbindPorts(
...
+        }
+    }
+	else {
 	    ret = vpp_acl_interface_unbind(hwif_name.c_str(), acl_swindex, is_input);
-
+        if(ret == 0) {
+            hwif_list.remove(hwif_name);
+            tunterm_acl_bindunbind(tbl_oid, false, hwif_name);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 4 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 5 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 6 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Move the ports_map unbind update above directly in the unbind case

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 7 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: In this case want to still break out of the loop when SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is false - updated the same check later on to be in a single 'if' statement.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 8 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.h (hunk 9 above)
Comment (dypet) [reply platform/saivpp/vpplib/vppxlate/SaiVppXlate.h]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 10 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 11 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: The check for acl->count > 0 was added here to skip this block for the cases where there are only tunterm acl rules and no standard acl rules

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 13 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: When iterating through the incoming rule's sai attributes we don't know if it's a tunterm acl or regular acl until SAI_ACL_ENTRY_ATTR_FIELD_TUNNEL_TERMINATED is encountered. There are common attr between the two, like SAI_ACL_ENTRY_ATTR_FIELD_DST_IP, which would need to be populated in both cases. Both rules (regular acl and tunterm acl) are updated in parallel while looping through the sai attributes, and then at the end once we know if it's a tunterm rule or not we clear the other rule which potentially had fields populated.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 12 above)
Comment (yue-fred-gao) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:798]: Conceptually with tunterm feature, each ACL table can have 1 or 2 ACL objects in VPP: one for regular ACL and one for tunterm ACL. In the create/update request, we go through the ACEs and convert them to 2 lists of vpp_acl_t and vpp_tunterm_acl_t (you can create a super data structure including the 2 lists). Then we pass the 2 lists along with table_id to a function. The function reads acl_swindex and tunterm_acl_swindex from table_id and does necessary operation to create/update/delete acl or tunterm acl from vpp.

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 20]
```diff
@@ -405,6 +405,148 @@ sai_status_t acl_rule_field_update(
...
+        vpp_ip_addr_t *ip_addr, *ip_mask;
+        ip_addr = &rule->dst_prefix;
+        ip_mask = &rule->dst_prefix_mask;
+        struct sockaddr_in *sin =  &ip_addr->addr.ip4;
+        ip_addr->sa_family = AF_INET;
+        sin->sin_addr.s_addr = value->aclfield.data.ip4;
+        sin =  &ip_mask->addr.ip4;
+        sin->sin_addr.s_addr = value->aclfield.mask.ip4;
+        vpp_ip_addr_t_to_string(ip_addr, dst_ip_str, INET6_ADDRSTRLEN);
+        SWSS_LOG_NOTICE("Tunterm acl rule has dst IP: %s", dst_ip_str);
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:522]: Please change all NOTICE to INFO except very few important one like tunterm acl created/modify/deleted successfully. See issue #116 .

File: platform/saivpp/vpplib/SwitchStateBase.h [hunk 21]
```diff
@@ -992,7 +992,9 @@ namespace saivpp
 
         private:
             std::map<sai_object_id_t, std::list<sai_object_id_t>> m_acl_tbl_rules_map;
+            std::map<sai_object_id_t, std::list<std::string>> m_acl_tbl_hw_ports_map;
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/SwitchStateBase.h:995]: There is already m_acl_tbl_grp_ports_map. If we see acl and tunterm_acl as two objects created from the same acl_table, it is possible to share the same mapping. even m_acl_swindex_map and m_tunterm_acl_swindex_map can be combined to one if we use std::pair or std::Array to keep both indexes in one

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c [hunk 22]
```diff
@@ -1927,6 +1989,144 @@ int vpp_acl_del (uint32_t acl_index)
...
+        } else if (in_rule->ip_protocol == 2) {
+            vpp_rule->path.proto  = htonl(FIB_API_PATH_NH_PROTO_IP6);
+            memcpy(vpp_rule->path.nh.address.ip6, &in_rule->next_hop_ip.addr.ip6.sin6_addr.s6_addr, sizeof(vpp_rule->path.nh.address.ip6));
+        } else {
+            SAIVPP_WARN("Unknown protocol in next hop prefix");
+            VPP_UNLOCK();
+            return -EINVAL;
+        }
+    }
+    mp->context = store_ptr(tunterm_index);
```
Comment (yue-fred-gao) [inline platform/saivpp/vpplib/vppxlate/SaiVppXlate.c:2119]: why do we need it in context? Is it already passed in with tunterm_acl_index?

File: platform/saivpp/vpplib/SwitchStateBase.h (hunk 14 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBase.h]: Added

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 15 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 23]
```diff
@@ -489,6 +663,94 @@ static bool cmp_priority (
...
+        m_tunterm_acl_swindex_map[tbl_oid] = tunterm_acl_swindex;
+    } else {
+        return SAI_STATUS_FAILURE;
+    }
+
+    if(do_port_bind) {
+        auto it_hw_ports = m_acl_tbl_hw_ports_map.find(tbl_oid);
+        if (it_hw_ports == m_acl_tbl_hw_ports_map.end()) {
+            auto sid = sai_serialize_object_id(tbl_oid);
+            SWSS_LOG_WARN("Tunterm acl - no ports bound for table id %s", sid.c_str());
```
[the remark this answers is not in the document]
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp [hunk 24]
```diff
@@ -778,6 +1081,8 @@ sai_status_t SwitchStateBase::AclTblRemove(
 	m_acl_tbl_rules_map.erase(it);
     }
 
+    status = tunterm_acl_delete(tbl_oid);
+
     auto vpp_idx_it = m_acl_swindex_map.find(tbl_oid);
```
[the remark this answers is not in the document]
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp:1188]: Regular ACL will still exist with default config and will need to be cleaned up

File: platform/saivpp/vpplib/SwitchStateBaseAcl.cpp (hunk 20 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBaseAcl.cpp]: Updated

File: platform/saivpp/vpplib/SwitchStateBase.h (hunk 21 above)
Comment (dypet) [reply platform/saivpp/vpplib/SwitchStateBase.h:996]: Will consider merging some of these mappings as a future enhancement

File: platform/saivpp/vpplib/vppxlate/SaiVppXlate.c (hunk 22 above)
Comment (dypet) [reply platform/saivpp/vpplib/vppxlate/SaiVppXlate.c:2119]: In create case the VPP tunterm sw_index will be returned here

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 25]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
+    uint32_t     tunterm_acl_swindex = 0;
+    bool         do_port_bind = false;
+    auto         tbl_sid = sai_serialize_object_id(tbl_oid);
+
+    auto tunterm_idx_it = m_tunterm_acl_swindex_map.find(tbl_oid);
+    if ((tunterm_idx_it == m_tunterm_acl_swindex_map.end()) && (acl != NULL)) {
+        tunterm_acl_swindex = ~tunterm_acl_swindex;
+        do_port_bind = true;
+    } else if (tunterm_idx_it != m_tunterm_acl_swindex_map.end() && (acl == NULL)) {
+        status = tunterm_acl_delete(tbl_oid, false);
```
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:928]: Just return the call. It makes it clear we are short-circuiting this function call here.

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 26]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
+}
 
-	    for (uint32_t i = 0; i < p_ace->attrs_count; i++) {
-		attr = &p_ace->attrs[i];
+sai_status_t SwitchStateBase::tunterm_acl_bindunbind(sai_object_id_t tbl_oid, bool is_add, std::string hwif_name)
+{
+    auto tunterm_idx_it = m_tunterm_acl_swindex_map.find(tbl_oid);
+    if (tunterm_idx_it != m_tunterm_acl_swindex_map.end()) {
+        auto tunterm_idx = tunterm_idx_it->second;
+        vpp_tunterm_acl_interface_add_del(tunterm_idx, is_add, hwif_name.c_str());
```
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:911]: Any reason you don't check status of `vpp_tunterm_acl_interface_add_del` or `tunterm_acl_bindunbind`?

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 27]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
-	}
+    status = vpp_tunterm_acl_add_replace(&tunterm_acl_swindex, acl->count, acl);
+
+    if (status == SAI_STATUS_SUCCESS) {
+        m_tunterm_acl_swindex_map[tbl_oid] = tunterm_acl_swindex;
+    } else {
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (do_port_bind) {
```
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:947]: Optional: add a comment here explaining the intent of this block, ie. "enable tunterm ACL on ports to which the table is already bound"

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 28]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
+        }
+    }
+
+    uint32_t tunterm_acl_swindex = tunterm_idx_it->second;
+    status = vpp_tunterm_acl_del(tunterm_acl_swindex);
+    if (status == SAI_STATUS_SUCCESS) {
+        m_tunterm_acl_swindex_map.erase(tunterm_idx_it);      
+    }
+    
+    if(table_delete && (status == SAI_STATUS_SUCCESS)) {
```
Comment (AkeelAli) [inline saivpp/src/SwitchStateBaseAcl.cpp:1000]: You've already checked STATUS_SUCCESS above, you can move this check on table_delete inside the same block or return on non success.

**Review by AkeelAli — APPROVED** [review]
Minor comments. We can add an HLD later to detail the feature + how to test it.

File: saivpp/src/vppxlate/SaiVppXlate.h [hunk 29]
```diff
@@ -81,6 +81,21 @@ extern "C" {
         vpp_acl_rule_t rules[0];
     } vpp_acl_t;
 
+    typedef struct {
+        vpp_ip_addr_t dst_prefix;
+        vpp_ip_addr_t dst_prefix_mask;
+        char hwif_name[64];
+        uint8_t  ip_protocol;
+        vpp_ip_addr_t next_hop_ip;
+    } tunterm_acl_rule_t;
```
[the remark this answers is not in the document]
Comment (dypet) [reply saivpp/src/vppxlate/SaiVppXlate.h]: Updated

File: saivpp/src/vppxlate/SaiVppXlate.h [hunk 30]
```diff
@@ -216,6 +231,10 @@ typedef enum {
 
     extern int vpp_acl_add_replace(vpp_acl_t *in_acl, uint32_t *acl_index, bool is_replace);
     extern int vpp_acl_del(uint32_t acl_index);
+    extern int vpp_tunterm_acl_interface_add_del (uint32_t tunterm_index,
+                                           bool is_bind, const char *hwif_name);
+    extern int vpp_tunterm_acl_del (uint32_t tunterm_index);
+    extern int vpp_tunterm_acl_add_replace (uint32_t *tunterm_index, uint32_t count, vpp_tunterm_acl_t *acl);
```
[the remark this answers is not in the document]
Comment (dypet) [reply saivpp/src/vppxlate/SaiVppXlate.h]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 31]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
 
-		if (attr->id == SAI_ACL_ENTRY_ATTR_ACTION_COUNTER) {
-		    ace_cntr_oid = attr->value.aclaction.parameter.oid;
+sai_status_t SwitchStateBase::tunterm_acl_add_replace(vpp_tunterm_acl_t *acl, sai_object_id_t tbl_oid)
+{
+    sai_status_t status = SAI_STATUS_SUCCESS;
+    uint32_t     tunterm_acl_swindex = 0;
+    bool         do_port_bind = false;
+    auto         tbl_sid = sai_serialize_object_id(tbl_oid);
+
```
[the remark this answers is not in the document]
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp:926]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 32]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
+    sai_object_id_t tbl_oid,
+    size_t &n_total_entries,
+    acl_tbl_entries_t *&aces,
+    std::list<ordered_ace_list_t> &ordered_aces)
 {
-    auto it = m_acl_tbl_rules_map.find(tbl_oid);
-
-    sai_status_t status = SAI_STATUS_SUCCESS;
+    sai_status_t                status = SAI_STATUS_SUCCESS;
+    auto it =                   m_acl_tbl_rules_map.find(tbl_oid);
```
[the remark this answers is not in the document]
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 33]
```diff
@@ -461,202 +604,458 @@ sai_status_t acl_priority_attr_get (
...
+        for (uint32_t i = 0; i < p_ace->attrs_count && status == SAI_STATUS_SUCCESS; i++) {
+            attr = &p_ace->attrs[i];
+            auto meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_ACL_ENTRY, attr->id);
+
+            if (meta != NULL) {
+                SWSS_LOG_INFO("Type %s attrib id %s",
+                    sai_serialize_object_type(SAI_OBJECT_TYPE_ACL_ENTRY).c_str(),
+                    meta->attridname);
+            }
+            if (attr->id == SAI_ACL_ENTRY_ATTR_FIELD_ACL_RANGE_TYPE && !ace.is_tunterm) {
```
[the remark this answers is not in the document]
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp (hunk 27 above)
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp:958]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp (hunk 28 above)
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp [hunk 34]
```diff
@@ -1274,11 +1683,20 @@ sai_status_t SwitchStateBase::aclBindUnbindPorts(
...
 
-	if (is_bind)
+	if (is_bind) {
 	    ret = vpp_acl_interface_bind(hwif_name.c_str(), acl_swindex, is_input);
-	else
+	    if (ret == 0) {
+	        tunterm_acl_bindunbind(tbl_oid, true, hwif_name);
+	        m_acl_tbl_hw_ports_map[tbl_oid].push_back(hwif_name);
+	    }
+	}
```
[the remark this answers is not in the document]
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

File: saivpp/src/SwitchStateBaseAcl.cpp (hunk 25 above)
Comment (dypet) [reply saivpp/src/SwitchStateBaseAcl.cpp]: Updated

**Pull request merged**